src/main/scala/org/kframework/kore/interfaces/Interfaces.scala

```diff
@@ -481,61 +481,183 @@ object pattern {
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Sentence
+  }
+
+  trait Import extends Sentence {
+    val name: p.Name
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Import
+  }
+
+  object Import {
+    def unapply(arg: Import): Option[(p.Name, Attributes)] = Some(arg.name, arg.att)
+  }
+
+  trait SortDeclaration extends Sentence {
+    val sort: p.Sort
+
+    override lazy val sorts = Set(sort)
+
+    override def onAttributes(f: p.Pattern => p.Pattern): SortDeclaration
+  }
+
+  object SortDeclaration {
+    def unapply(arg: SortDeclaration): Option[(p.Sort, Attributes)] = Some(arg.sort, arg.att)
+  }
+
+  trait SymbolDeclaration extends Sentence {
+    val sort: p.Sort
+    val symbol: p.Symbol
+    val args: Seq[p.Sort]
+
+    // TODO: Should `sorts` include the `args` here as well?
+    override lazy val sorts = Set(sort)
+    override lazy val symbols = Set(symbol)
+
+    override def onAttributes(f: p.Pattern => p.Pattern): SymbolDeclaration
+  }
+
+  object SymbolDeclaration {
+    def unapply(arg: SymbolDeclaration): Option[(p.Sort, p.Symbol, Seq[p.Sort], Attributes)] = Some(arg.sort, arg.symbol, arg.args, arg.att)
+  }
+
+  trait Rule extends Sentence {
+    val pattern: p.Pattern
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Rule
+  }
+
+  object Rule {
+    def unapply(arg: Rule): Option[(p.Pattern, Attributes)] = Some(arg.pattern, arg.att)
+  }
+
+  trait Axiom extends Sentence {
+    val pattern: p.Pattern
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Axiom
+  }
+
+  object Axiom {
+    def unapply(arg: Axiom): Option[(p.Pattern, Attributes)] = Some(arg.pattern, arg.att)
   }
 
 }
 
+object builders {
+
+  trait OuterBuilders {
+    import outer._
+    import pattern.{Name, Sort, Symbol, Pattern}
+
+    def Definition(modules: Seq[Module], att: Attributes): Definition
+    def Module(name: Name, sentences: Seq[Sentence], att: Attributes): Module
+    def Import(name: Name, att: Attributes): Import
+    def SortDeclaration(sort: Sort, att: Attributes): SortDeclaration
+    def SymbolDeclaration(sort: Sort, symbol: Symbol, args: Seq[Sort], att: Attributes): SymbolDeclaration
+    def Rule(pattern: Pattern, att: Attributes): Rule
+    def Axiom(pattern: Pattern, att: Attributes): Axiom
+  }
+
+  trait PatternBuilders {
+    import pattern._
+
+    def Variable(_1: Name, _2: Sort): Variable
+    def DomainValue(_1: Symbol, _2: Value): DomainValue
+    def Top(): Top
+    def Bottom(): Bottom
+    def Not(_1: Pattern): Not
+    def Next(_1: Pattern): Next
+    def And(_1: Pattern, _2: Pattern): And
+    def Or(_1: Pattern, _2: Pattern): Or
+    def Implies(_1: Pattern, _2: Pattern): Implies
+    def Equals(_1: Pattern, _2: Pattern): Equals
+    def Exists(_1: Variable, _2: Pattern): Exists
+    def ForAll(_1: Variable, _2: Pattern): ForAll
+    def Rewrite(_1: Pattern, _2: Pattern): Rewrite
+    def Application(_1: Symbol, args: Seq[Pattern]): Application
+  }
+
```

Comment:
@daejunpark, if we merge them, we shoot ourselves in the foot regarding being able to create a definition using default pattern builders, then be able to create specialized pattern builders from it, something I know you agree of.